### PR TITLE
feat: blocks turn invisible

### LINKED ISSUE #42: blocks turn invisible when between player and camera

### PR DESCRIPTION
**Pull request by Rinder5**
fixes #42

### CODE REVIEW

**Comment by nicholaschuayunzhi** [comment]
@Rinder5 Just some first impressions so far. I noticed that the blocks fadeout logic is abit inconsistent. Does it depend on the player position as well?

image

File: Assets/Scripts/BlockBehaviour.cs [hunk 1]
```diff
@@ -18,17 +18,26 @@ public class BlockBehaviour : MonoBehaviour
     private bool _isPlayerStandingOn;
     private BlockFace _lastClickedFace;
     private BlockFaceBehaviour _blockFaceBehaviour;
+    private Renderer _renderer;
+    private bool _isFading = false;
```
Comment (nicholaschuayunzhi) [inline Assets/Scripts/BlockBehaviour.cs]: Not too comfortable having this in `BlockBehaviour`. Not all blocks will be "wall" blocks. Having_renderer is fine though.

File: Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    // Update is called once per frame
+    public override void OnUpdate()
+    {
+	    Renderer renderer = _block.GetRenderer();
+        if (renderer.material.color.a > FadeMin && _block.IsFading())
+	        renderer.material.color = new Color(renderer.material.color.r, renderer.material.color.g, renderer.material.color.b, renderer.material.color.a - FadeSpeed);
+
+        if (renderer.material.color.a < FadeMax && !_block.IsFading())
+            renderer.material.color = new Color(renderer.material.color.r, renderer.material.color.g, renderer.material.color.b, renderer.material.color.a + FadeSpeed);
```
Comment (nicholaschuayunzhi) [inline Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs]: We could get a reference to the camera from here. 1 Possibility would be to get a direct reference to the camera script. We may need multiple references to camera to account for different cameras across stages. That way,we inject the dependency via the inspector and not the script.

Comment (nicholaschuayunzhi) [reply Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs]: Another thing you could do is get the camera based on tag. But could be costly on start.

File: Assets/Scripts/CameraController.cs [hunk 3]
```diff
@@ -59,12 +68,30 @@ private void Update()
...
+        Vector3 cameraPosition = transform.position;
+        Vector3 playerPosition = PlayerObject.transform.position;
+        foreach (BlockBehaviour blockBehavior in _wallBlocks)
+        {
+            if ((cameraPosition - blockBehavior.transform.position).magnitude > 
+                (cameraPosition - playerPosition).magnitude)
+                blockBehavior.SetIsFading(false);
+            else
+                blockBehavior.SetIsFading(true);
+        }
```
Comment (nicholaschuayunzhi) [inline Assets/Scripts/CameraController.cs]: Could use delegates to execute functions in FadeoutPlugins. Or could just directly set is fading in the FadeoutPlugin

**Review by nicholaschuayunzhi — CHANGES_REQUESTED** [review]
@Rinder5 left some comments. The blocks turning invisible does not feel very intuitive. We may have to discuss with team again.

File: Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs (hunk 2 above)
Comment (Rinder5) [reply Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs]: I think getting camera based on tag is a good idea. Not much different from getting blocks from tag I think.

**Comment by Rinder5** [comment]
Yes, also based on player position. The decision checking is pretty simple. Basically we just draw a direct line from the camera to the block and the camera to the player. If the camera to the block is shorter, we turn it invisible.

File: Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs (hunk 2 above)
Comment (nicholaschuayunzhi) [reply Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs]: Ok we can try that. I'm just concerned about costs. Do note, i believe you must use _block's monobehaviour to get the object by tag.

Comment (Rinder5) [reply Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs]: I think the plugin itself already has details about the block? If so, we can pass in the block object to the camera on startup

File: ProjectSettings/TagManager.asset [hunk 4]
```diff
@@ -3,7 +3,8 @@
 --- !u!78 &1
 TagManager:
   serializedVersion: 2
-  tags: []
+  tags:
+  - Wall
```
Comment (nicholaschuayunzhi) [inline ProjectSettings/TagManager.asset]: can we remove this

File: Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+using UnityEngine;
+using UnityEngine.Rendering;
+
+[CreateAssetMenu]
+public class FadeoutPlugin : BlockPlugin
+{
+    public const float FadeSpeed = 0.1f;
+    public const float FadeMin = 0.0f;
+    public const float FadeMax = 1.0f;
+    public Renderer Renderer;
```
Comment (nicholaschuayunzhi) [inline Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs]: Could we have this as a private variable `_renderer`

File: Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    public bool IsFading = false;
+
+    public override void Plug(BlockBehaviour blockBehaviour)
+    {
+        Renderer = blockBehaviour.GetComponent<Renderer>();
+        Renderer.shadowCastingMode = ShadowCastingMode.Off;
+        CameraController cameraController = Camera.main.GetComponent<CameraController>();
+        _block = blockBehaviour;
```
Comment (nicholaschuayunzhi) [inline Assets/ScriptableObjects/Plugin Scripts/FadeoutPlugin.cs]: could we call the super class method's `Plug(BlockBehaviour)` at line 16?

**Comment by Rinder5** [comment]
@nicholaschuayunzhi thanks for all your helpful comments! I made the requested changes

**Review by nicholaschuayunzhi — APPROVED** [review]
LGTM

**Pull request merged**